File: src/main/java/com/marsprobe/commandcenter/entities/Field.java
package com.marsprobe.commandcenter.entities;

import java.util.List;

public class Field {

	private int id;
	private int limitX;
	private int limitY;
	private List<Probe> probes;
	
	public Field(int id, int limitX, int limitY, List<Probe> probes) {
		super();
		this.id = id;
		this.limitX = limitX;
		this.limitY = limitY;
		this.probes = probes;
	}

	public Field() {
		super();
	}

	public int getId() {
		return id;
	}

	public void setId(int id) {
		this.id = id;
	}

	public int getLimitX() {
		return limitX;
	}

	public void setLimitX(int limitX) {
		this.limitX = limitX;
	}

	public int getLimitY() {
		return limitY;
	}

	public void setLimitY(int limitY) {
		this.limitY = limitY;
	}

	public List<Probe> getProbes() {
		return probes;
	}

	public void setProbes(List<Probe> probes) {
		this.probes = probes;
	}

	@Override
	public String toString() {
		return "Field [id=" + id + ", limitX=" + limitX + ", limitY=" + limitY + ", probes=" + probes + "]";
	}

}

File: src/main/kotlin/com/elo7/commandcenter/field/Field.kt
package com.elo7.commandcenter.field

import com.elo7.commandcenter.vehicle.Position
import com.elo7.commandcenter.vehicle.Vehicle

class Field(
    private val id: Int,
    private val name: String,
    private val limitX: Int,
    private val limitY: Int,
    private val vehicles: MutableList<Vehicle>
) {
    fun getLimitX() = limitX
    fun getLimitY() = limitY
    fun getVehicles() = vehicles

    fun saveVehicle(request: Vehicle) : Field {
        return when (getVehicleById(request.getVehicleId()) != null) {
            true -> {
                if (isPositionAvailable(request.getVehicleId(), request.getCurrentPosition()))
                    vehicles[vehicles.indexOf(getVehicleById(request.getVehicleId()))] = request
                Field(id, name, limitX, limitY, vehicles)
            }
            else -> {
                vehicles.add(request)
                Field(id, name, limitX, limitY, vehicles)
            }
        }
    }

    private fun getVehicleById(id: Int) : Vehicle? {
        return vehicles.find { it.getVehicleId() == id }
    }

    fun isPositionAvailable(vehicleId: Int, position: Position): Boolean {
        val occupiedPositions = vehicles.associate { it.getVehicleId() to it.getCurrentPosition() }
        return when {
            occupiedPositions.containsValue(position) ->
                vehicleId == occupiedPositions.filterValues { it == position }.keys.first()
            else -> true
        }
    }

    override fun toString(): String {
        return super.toString()
    }
}